Rename abstract Authenticate => HandleAuthenticate

HaoK opened the pull request:
Also:
- fix ShouldHandleScheme to only check for string.Empty
- use the Once version of Authenticate more consistently

Fixes https://github.com/aspnet/Security/issues/341, https://github.com/aspnet/Security/issues/338

cc @Tratcher

HaoK commented:
Add a few tests

Tratcher commented on src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs:
```diff
@@ -161,7 +161,7 @@ protected Task<AuthenticationTicket> AuthenticateOnceAsync()
             if (_authenticateTask == null)
             {
-                _authenticateTask = AuthenticateAsync();
+                _authenticateTask = HandleAuthenticateAsync();
```
It's odd that you renamed AuthenticateAsync -> HandleAuthenticateAsync, but left AuthenticateOnceAsync as is. HandleAuthenticateOnceAsync?

HaoK replied:
Fair enough I'll rename that too

Tratcher commented on src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs:
```diff
@@ -189,12 +189,12 @@ protected Task<AuthenticationTicket> AuthenticateOnceAsync()
         {
             return string.Equals(BaseOptions.AuthenticationScheme, authenticationScheme, StringComparison.Ordinal) ||
-                (BaseOptions.AutomaticAuthentication && string.IsNullOrWhiteSpace(authenticationScheme));
+                (BaseOptions.AutomaticAuthentication && authenticationScheme == string.Empty);
```
Use IsNullOrEmpty. Never use == with strings.

HaoK replied:
Brent brought up an issue that we use automatic with an empty string never null. I am on the fence. 

But I am fine with treating null and empty the same as we were before if you feel strongly about it.

Tratcher replied:
I could have sworn there was a dot net design guideline for treating string.Empty and null as the same, but I can't find a reference for it.

HaoK replied:
Well I don't like treating them differently in general, but since we will be using [NotNull] for authenticationScheme everywhere, I don't care too strongly about this.  You have the tiebreaking vote :)

Tratcher commented:
:shipit:

HaoK commented:
Updated PR to fix cookie issues since this was touching the HandleAuthenticate code

@Tratcher can you review again, updated Cookie and TicketSerializer (to prevent nulls)

HaoK commented:
Fixes https://github.com/aspnet/Security/issues/355 and https://github.com/aspnet/Security/issues/325 now as well

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -28,74 +28,88 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
         private string _sessionKey;
 
-        protected override async Task<AuthenticationTicket> AuthenticateAsync()
+        private async Task<AuthenticationTicket> ConfigureTicket()
```
RetrieveTicketAsync? Also, is there any point in having this run more than once? RetrieveTicketOnceAsync?

HaoK replied:
Went with EnsureCookieTicket read and stored in a field

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -299,7 +314,7 @@ private CookieOptions BuildCookieOptions()
             {
                 var exceptionContext = new CookieExceptionContext(Context, Options,
-                    CookieExceptionContext.ExceptionLocation.ApplyResponseGrant, exception, model);
+                    CookieExceptionContext.ExceptionLocation.ApplyResponseGrant, exception, ticket);
```
These ExceptionLocations are out of date.

HaoK replied:
Added enums for each virtual (SignIn, SignOut, Unauthorized, Forbidden, FinishResponse) this seems a bit much to me...

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -183,7 +197,7 @@ private CookieOptions BuildCookieOptions()
+            var ticket = await HandleAuthenticateOnceAsync();
             try
             {
                 await ApplyCookie(ticket);
```
This is the only place that ApplyCookie is called. Consider inlining it or refactoring it so that it can also be called from HandleSignIn

HaoK replied:
I'll see if I can combine it with the sign in one

HaoK replied:
Moved to a ApplyHeaders(bool shouldRedirectToReturnUrl) helper

Tratcher commented:
:watch:

HaoK commented:
PR updated

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -27,75 +27,103 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
+                    // Finally we have a valid ticket
+                    _cookieTicket = ticket;
+                }
+                finally
```
drop the try/finally and just set _cookieRead at the top.

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -27,75 +27,103 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
+            try
+            {
+                await EnsureCookieTicket();
+                ticket = _cookieTicket;
```
Did you notice that every usage of `await EnsureCookieTicket();` is immediately followed by `ticket = _cookieTicket;`? Why not use the same pattern as HandleAuthenticateOnceAsync?

HaoK replied:
You mean the pattern in the base class?  Its for the exception handler which sometimes changes the ticket... for some of these methods...  Yes I noticed this code looks bad :)  Let me take a look again

HaoK replied:
Alright cleaned things up to use the pattern I used in AuthHandler

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -27,75 +27,102 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
         private DateTimeOffset? _renewExpiresUtc;
         private string _sessionKey;
+        private Task<AuthenticationTicket> _cookieTicketTask;
+        private bool _cookieRead;
```
_cookieRead is no longer used.

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -27,75 +27,102 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
+            }
+            return _cookieTicketTask;
+        }
+
```
nit: extra line

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -270,36 +289,13 @@ private CookieOptions BuildCookieOptions()
-                    HeaderNameExpires,
-                    HeaderValueMinusOne);
-
                 var shouldLoginRedirect = Options.LoginPath.HasValue && Request.Path == Options.LoginPath;
```
Request.Path may be wrong (altered by Map) when SignIn is applied immediately. Do we need to add base.OriginalPath like we have for OriginalPathBase?

HaoK replied:
Yeah looks like it would be the same issue, I'll add a test and fix if its busted

Tratcher commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -387,36 +377,34 @@ private static bool IsHostRelative(string path)
+                    Request.Scheme +
+                    "://" +
+                    Request.Host +
+                    Request.PathBase +
```
Same base.OriginalPathBase issue here.

Tratcher replied:
And in HandleUnauthorizedAsync

HaoK commented:
Updated with new tests and fixed signin/signout to reflect map correctly.  I'll push this so I can start on the API minor changes, and address any further PR cleanup in the next PR

Tratcher commented:
:shipit:

HaoK commented:
ab4ba794e546074a573fbf76d9150ffcf9752c89
